crates/memory/src/op_store/time_slice_hash_store.rs

```diff
@@ -0,0 +1,532 @@
+                    }
+                    highest_stored_id = *id;
+                }
+                SliceHash::Block { id, count, .. } => {
+                    if *id != previous_end.wrapping_add(1) {
+                        println!(
+                            "Block does not follow previous: {} != {}",
+                            *id,
+                            previous_end.wrapping_add(1)
+                        );
+                        return false;
+                    }
+
+                    if *count > u64::MAX - *id {
+                        // Not allowed to wrap around
+                        println!("Block wraps around");
+                        return false;
+                    }
+                }
+            }
+        }
+
+        if highest_stored_id != 0
+            && highest_stored_id != self.highest_stored_id.unwrap()
+        {
+            println!(
+                "highest_stored_id does not match: {} != {:?}",
+                highest_stored_id, self.highest_stored_id
+            );
+            return false;
+        }
+
+        match self.inner.last() {
+            Some(SliceHash::Block { id, count }) => {
+                if *count != u64::MAX - id {
+                    println!(
+                        "Block does not cover remaining range: {} != {}",
+                        *count,
+                        u64::MAX - highest_stored_id
+                    );
+                    return false;
+                }
+            }
+            Some(SliceHash::Single { id, .. }) => {
+                if *id != highest_stored_id || *id != u64::MAX {
+                    println!(
+                        "Single does not cover remaining range: {} != {}",
+                        *id, highest_stored_id
+                    );
+                    return false;
+                }
+            }
+            _ => {
+                println!("No last element");
+                return false;
+            }
+        }
+
+        true
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn create_empty() {
+        let store = TimeSliceHashStore::new();
+
+        assert_eq!(None, store.highest_stored_id);
+        assert_eq!(1, store.inner.len());
+        assert_eq!(
+            SliceHash::Block {
+                id: 0,
+                count: u64::MAX
+            },
+            store.inner[0]
+        );
+
+        assert!(store.check());
+    }
+
+    #[test]
+    fn insert_empty_hash_into_empty() {
+        let mut store = TimeSliceHashStore::new();
+        let original = store.clone();
+
+        store.insert(100, bytes::Bytes::new()).unwrap();
+
+        // Doesn't change the stored value, but does update the highest stored value
+        assert_eq!(original.inner, store.inner);
+        assert_eq!(None, store.highest_stored_id);
+
+        assert!(store.check());
+    }
+
+    #[test]
+    fn insert_single_hash_into_empty() {
```

Comment:
maybe one more test `insert_single_hash_into_block` where you explicitly create the first block, even though i know its covered here